packages/@react-aria/dnd/src/useClipboard.ts

```diff
@@ -0,0 +1,139 @@
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import {chain} from '@react-aria/utils';
+import {DragItem, DropItem} from '@react-types/shared';
+import {HTMLAttributes, useEffect, useRef} from 'react';
+import {readFromDataTransfer, writeToDataTransfer} from './utils';
+import {useFocus} from '@react-aria/interactions';
+
+interface ClipboardProps {
+  getItems?: () => DragItem[],
+  onCopy?: () => void,
+  onCut?: () => void,
+  onPaste?: (items: DropItem[]) => void
+}
+
+interface ClipboardResult {
+  clipboardProps: HTMLAttributes<HTMLElement>
+}
+
+const globalEvents = new Map();
+function addGlobalEventListener(event, fn) {
+  let eventData = globalEvents.get(event);
+  if (!eventData) {
+    let handlers = new Set<(e: Event) => void>();
+    let listener = (e) => {
+      for (let handler of handlers) {
+        handler(e);
+      }
+    };
+
+    eventData = {listener, handlers};
+    globalEvents.set(event, eventData);
+
+    document.addEventListener(event, listener);
+  }
+
+  eventData.handlers.add(fn);
+  return () => {
+    eventData.handlers.delete(fn);
+    if (eventData.handlers.size === 0) {
+      document.removeEventListener(event, eventData.listener);
+      globalEvents.delete(event);
+    }
+  };
+}
+
+export function useClipboard(options: ClipboardProps): ClipboardResult {
+  let ref = useRef(options);
+  ref.current = options;
+
+  let isFocusedRef = useRef(false);
+  let {focusProps} = useFocus({
+    onFocusChange: (isFocused) => {
+      isFocusedRef.current = isFocused;
+    }
+  });
+
+  useEffect(() => {
+    let onBeforeCopy = (e: ClipboardEvent) => {
+      // Enable the "Copy" menu item in Safari if this element is focused and copying is supported.
+      let options = ref.current;
+      if (isFocusedRef.current && options.getItems) {
+        e.preventDefault();
+      }
+    };
+
+    let onCopy = (e: ClipboardEvent) => {
+      let options = ref.current;
+      if (!isFocusedRef.current || !options.getItems) {
+        return;
+      }
+
+      e.preventDefault();
+      writeToDataTransfer(e.clipboardData, options.getItems());
+      options.onCopy?.();
+    };
+
+    let onBeforeCut = (e: ClipboardEvent) => {
+      let options = ref.current;
+      if (isFocusedRef.current && options.onCut && options.getItems) {
+        e.preventDefault();
+      }
+    };
+
+    let onCut = (e: ClipboardEvent) => {
+      let options = ref.current;
+      if (!isFocusedRef.current || !options.onCut || !options.getItems) {
+        return;
+      }
+
+      e.preventDefault();
+      writeToDataTransfer(e.clipboardData, options.getItems());
+      options.onCut();
```

Comment:
I noticed in the story example that the cut operation removed the item from the list immediately since the `list.remove(...keys)` is called in the `onCut` callback. If a user wanted the item to only be removed once the paste happens (like how the file system works on windows), would we expect that they would need to track if the last operation was a copy/cut themselves or perhaps we could track that internally here and pass it through to the `onPaste` callback?

Reply:
Huh that's weird, I never realized Windows behaved like that. Cut doesn't behave like that for any other type of object I don't think? e.g. cutting text always removes it immediately.

I tested on macOS and it seems you can't cut files at all, only copy. Then you can press Cmd + Option + V to move the file instead of copying it.

Anyway, I don't think it's really possible. In the dnd API there's an ondragend event that gets called on the source to tell you that the dragged items were moved or copied. But there's nothing similar for copy and paste, i.e. no way to know something was pasted. I guess it would be up to the user to handle this on paste, and decide whether they should move or copy the item.